[otp_ctrl/doc] Zeroization documentation (#28030)
lowRISC/opentitan: hw/ip_templates/otp_ctrl/doc/programmers_guide.md

It should also be noted that the effect of locking a partition via the digest only takes effect **after** the next system reset.
To prevent integrity check failures SW must therefore ensure that no more programming operations are issued to the affected partition after initiating the digest calculation sequence.

## Zeroization Sequence

> Zeroizing a partition is completely handled by firmware and resembles an ordinary in-field provisioning flow.

Reviewer: Could we tweak the 'completely handled by firmware' section here? From the perspective of the OTP Controller DAI interface, who drives it / how it is driven is really out of scope. Also, I kind of dislike 'firmware' as a catch-all term.
How about this?
```suggestion
A partition zeroization sequence is driven by accesses through the OTP Controller's DAI interface, and resembles an ordinary in-field provisioning flow.
```

1. The DAI has a dedicated `ZEROIZE` command through which a word (either 32-bit for software fuses or 64-bit for hardware fuses including the digest and zeroization fields) in a partition can be zeroized.
The entire address space of a partition is zeroizable which contrasts with the other DAI commands.
For example, the digest field is never writable through the `WRITE` command in a hardware partition.

    > Side effect: The OTP controller detects the first _successful_ word zeroization and disables periodic consistency checks for the corresponding partition as these can potentially fail when interrupting an ongoing zeroization procedure.
    Integrity checks for hardware partitions can proceed normally until the next reset as they only act on buffered data.

>     Although the fuses of a partition can be zeroized in any order, it is recommended to first zeroize the zeroization field at the end of a partition to mark it as zeroized or being zeroized in case the flow is interrupted and needs to be resumed at a later point.

Reviewer: Can we tweak this sentence? Too many 'zeroize's? :)

It should also be noted that the effect of locking a partition via the digest only takes effect **after** the next system reset.
To prevent integrity check failures SW must therefore ensure that no more programming operations are issued to the affected partition after initiating the digest calculation sequence.

## Zeroization Sequence

Zeroizing a partition is completely handled by firmware and resembles an ordinary in-field provisioning flow.

1. The DAI has a dedicated `ZEROIZE` command through which a word (either 32-bit for software fuses or 64-bit for hardware fuses including the digest and zeroization fields) in a partition can be zeroized.
The entire address space of a partition is zeroizable which contrasts with the other DAI commands.
For example, the digest field is never writable through the `WRITE` command in a hardware partition.

>     > Side effect: The `fuse_ctrl` detects the first _successful_ word zeroization and disables periodic consistency checks for the corresponding partition as these can potentially fail when interrupting an ongoing zeroization procedure.

Reviewer 2: ```suggestion
    > Side effect: The `otp_ctrl` detects the first _successful_ word zeroization and disables periodic consistency checks for the corresponding partition as these can potentially fail when interrupting an ongoing zeroization procedure.
```

    Integrity checks for hardware partitions can proceed normally until the next reset as they only act on buffered data.

    Although the fuses of a partition can be zeroized in any order, it is recommended to first zeroize the zeroization field at the end of a partition to mark it as zeroized or being zeroized in the case the flow is interrupted and needs to be resumed at a later point.

    > Side effect: A fuse macro usually signals an error if a write attempts to clear an already set data or ECC bit.

>     > Such an unintended bit flip can occur in the ECC part of a word during a zeroization (when only the data part is zeroized), hence the `ZEROIZE` command disables ECC when zeroizing a word, setting all bits in both the data and ECC part of a fuse word.

Reviewer 3: I like this inline "Side effect: ..." formatting using `>`. Very nice looking in the rendered preview of my editor.

2. A successful zeroization of a fuse word results in the number of set bits in the zeroized workd being returned to software in the `DIRECT_ACCESS_RDATA` registers for inspection bypassing the
descrambling mechanism if the word belongs to a secret partition.
This is the only way firmware can confirm, in the absence of a malicious tampering attempt, whether a fuse has been cleared.
The `ZEROIZE` command is idempotent, i.e., it can be retried multiple times such that a zeroization of an already zeroized word has no effect.

> 3. When firmware determines that a partition has been sufficiently zeroized, it should reset the `fuse_ctrl` such that the zeroized data is also reflected in the buffer registers.

Reviewer 2: ```suggestion
3. When firmware determines that a partition has been sufficiently zeroized, it should reset the `otp_ctrl` such that the zeroized data is also reflected in the buffer registers.
```

>      > Side effect: Upon initialization, the `otp_ctrl` will first read the zeroization field of a partition to determine whether it is in a zeroized state.

Reviewer 3: Nit: "`otp_ctrl`" / "OTP controller" as above

>      > Side effect: Upon initialization, the `fuse_ctrl` will first read the zeroization field of a partition to determine whether it is in a zeroized state.

Reviewer 2: ```suggestion
     > Side effect: Upon initialization, the `otp_ctrl` will first read the zeroization field of a partition to determine whether it is in a zeroized state.
```

For example, the digest field is never writable through the `WRITE` command in a hardware partition.

    > Side effect: The OTP controller detects the first _successful_ word zeroization and disables periodic consistency checks for the corresponding partition as these can potentially fail when interrupting an ongoing zeroization procedure.
    Integrity checks for hardware partitions can proceed normally until the next reset as they only act on buffered data.

    Although the fuses of a partition can be zeroized in any order, it is recommended to first erase the zeroization marker at the end of a partition to mark it as zeroized or in the process of being so in case the flow is interrupted and needs to be resumed at a later point.

    > Side effect: A fuse macro usually signals an error if a write attempts to clear an already set data or ECC bit.
    > Such an unintended bit flip can occur in the ECC part of a word during a zeroization (when only the data part is zeroized), hence the `ZEROIZE` command disables ECC when zeroizing a word, setting all bits in both the data and ECC part of a fuse word.

> 2. A successful zeroization of a fuse word results in the number of set bits in the zeroized work being returned to software in the `DIRECT_ACCESS_RDATA` registers bypassing the descrambling mechanism if the word belongs to a secret partition.

Author: Fixed.
